extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt

```diff
@@ -467,38 +501,81 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
 
   /**
    * Add a Feature to the GeojsonSource.
+   * If [onDataParsed] is provided and not null - data will be loaded in async mode.
+   * Otherwise method will be synchronous.
    *
-   * @param value the feature
+   * @param value the feature collection
+   * @param onDataParsed optional callback notifying when data is parsed on a worker thread
    */
-  fun feature(value: Feature) = apply {
-    setProperty(PropertyValue("data", value.toValue()))
-  }
+  fun feature(
+    value: Feature,
+    onDataParsed: ((GeoJsonSource) -> Unit)? = null
+  ) = applyGeoJsonData(value, onDataParsed)
 
   /**
    * Add a FeatureCollection to the GeojsonSource.
+   * If [onDataParsed] is provided and not null - data will be loaded in async mode.
+   * Otherwise method will be synchronous.
    *
    * @param value the feature collection
+   * @param onDataParsed optional callback notifying when data is parsed on a worker thread
    */
-  fun featureCollection(value: FeatureCollection) = apply {
-    setProperty(PropertyValue("data", value.toValue()))
-  }
+  fun featureCollection(
+    value: FeatureCollection,
+    onDataParsed: ((GeoJsonSource) -> Unit)? = null
+  ) = applyGeoJsonData(value, onDataParsed)
 
   /**
    * Add a Geometry to the GeojsonSource.
+   * If [onDataParsed] is provided and not null - data will be loaded in async mode.
+   * Otherwise method will be synchronous.
    *
-   * @param value the geometry
+   * @param value the feature collection
+   * @param onDataParsed optional callback notifying when data is parsed on a worker thread
    */
-  fun geometry(value: Geometry) = apply {
-    setProperty(PropertyValue("data", value.toValue()))
+  fun geometry(
+    value: Geometry,
+    onDataParsed: ((GeoJsonSource) -> Unit)? = null
+  ) = applyGeoJsonData(value, onDataParsed)
+
+  private fun GeoJson.toPropertyValue() = PropertyValue(
+    "data",
+    when (this) {
+      is Geometry -> toValue()
+      is FeatureCollection -> toValue()
+      is Feature -> toValue()
+      else -> RuntimeException("GeoJson data must be Geometry, FeatureCollection or Feature!")
+    }
+  )
+
+  private fun applyGeoJsonData(
+    data: GeoJson,
+    onDataParsed: ((GeoJsonSource) -> Unit)?
+  ): GeoJsonSource = apply {
+    onDataParsed?.let { listener ->
+      workerHandler.post {
+        val property = data.toPropertyValue()
+        mainHandler.post {
+          setProperty(property)
+          listener.invoke(this)
+        }
+      }
+    } ?: setProperty(data.toPropertyValue())
```

Comment:
if no `onDataParsed` is passed, does it mean we use synchronized geojson parsing?

Reply:
Yes, it is stated in docs above.
Please also keep in mind that synchronous mode is available here if updating `GeoJsonSource` (kinda supporting back compatibility) but setting via `GeoJsonSource.Builder` will be always async.

Reply:
:+1: eventhough majority of users would want async, I have heard about customers wanting to have the non-async method.